File: Topological sort - GFG/topological-sort.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution
{
	public:
	void dfs(int v, vector<int>* adj, vector<bool>& visited, stack<int>& st) {
        visited[v] = true;
        
        vector<int>& neighbors = adj[v];
        for (int neighbor : neighbors) {
            if (!visited[neighbor]) {
                dfs(neighbor, adj, visited, st);
            }
        }
        
        st.push(v);
    }
	vector<int> topoSort(int v, vector<int> adj[]) 
	{
	    vector<bool> visited(v, false);
        stack<int> st;
        
        for (int i = 0; i < v; i++) {
            if (!visited[i]) {
                dfs(i, adj, visited, st);
            }
        }
        
        vector<int> topologicalSort;
        while (!st.empty()) {
            topologicalSort.push_back(st.top());
            st.pop();
        }
        
        return topologicalSort;
    }
};

//{ Driver Code Starts.

/*  Function to check if elements returned by user
*   contains the elements in topological sorted form
*   V: number of vertices
*   *res: array containing elements in topological sorted form
*   adj[]: graph input
*/
int check(int V, vector <int> &res, vector<int> adj[]) {
    
    if(V!=res.size())
    return 0;
    
    vector<int> map(V, -1);
    for (int i = 0; i < V; i++) {
        map[res[i]] = i;
    }
    for (int i = 0; i < V; i++) {
        for (int v : adj[i]) {
            if (map[i] > map[v]) return 0;
        }
    }
    return 1;
}

int main() {
    int T;
    cin >> T;
    while (T--) {
        int N, E;
        cin >> E >> N;
        int u, v;

        vector<int> adj[N];

        for (int i = 0; i < E; i++) {
            cin >> u >> v;
            adj[u].push_back(v);
        }
        
        Solution obj;
        vector <int> res = obj.topoSort(N, adj);

        cout << check(N, res, adj) << endl;
    }
    
    return 0;
}
// } Driver Code Ends

File: Topological sort - GFG/topological-sort.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader read =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());

        while (t-- > 0) {
            ArrayList<ArrayList<Integer>> list = new ArrayList<>();
            String st[] = read.readLine().trim().split("\\s+");
            int edg = Integer.parseInt(st[0]);
            int nov = Integer.parseInt(st[1]);

            for (int i = 0; i < nov + 1; i++)
                list.add(i, new ArrayList<Integer>());

            int p = 0;
            for (int i = 1; i <= edg; i++) {
                String s[] = read.readLine().trim().split("\\s+");
                int u = Integer.parseInt(s[0]);
                int v = Integer.parseInt(s[1]);
                list.get(u).add(v);
            }

            int[] res = new Solution().topoSort(nov, list);

            if (check(list, nov, res) == true)
                System.out.println("1");
            else
                System.out.println("0");
        }
    }
    static boolean check(ArrayList<ArrayList<Integer>> list, int V, int[] res) {
        
        if(V!=res.length)
        return false;
        
        int[] map = new int[V];
        for (int i = 0; i < V; i++) {
            map[res[i]] = i;
        }
        for (int i = 0; i < V; i++) {
            for (int v : list.get(i)) {
                if (map[i] > map[v]) return false;
            }
        }
        return true;
    }
}

// } Driver Code Ends


/*Complete the function below*/


class Solution
{
    static int[] topoSort(int v, ArrayList<ArrayList<Integer>> adj) 
    {
        int inDegree[] = new int[v];
        for(ArrayList<Integer> list: adj){
            for(int element: list){
                inDegree[element]++;
            }
        }
        ArrayList<Integer> ans = new ArrayList<>();
        bfs(v, adj, inDegree, ans);
        int topologicalSort[] = new int[v];
        int i=0;
        for(int element: ans){
            topologicalSort[i++]=element;
        }
        return topologicalSort;
    }
    public static void bfs(int v, ArrayList<ArrayList<Integer>> adj, int inDegree[], ArrayList<Integer> ans){
        Queue<Integer> q = new LinkedList<>();
        for(int i=0;i<v;i++){
            if(inDegree[i]==0){
                q.add(i);
            }
        }
        while(!q.isEmpty()){
            int curr = q.poll();
            ans.add(curr);
            ArrayList<Integer> neighbors = adj.get(curr);
            for(int neighbor: neighbors){
                if(--inDegree[neighbor]==0){
                    q.add(neighbor);
                }
            }
        }
    }
}
